### ISSUE TITLE
Distinguish between "masking" and "enforcing impenetrability" in immersed boundary routines

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#2249]
Is this correct?

https://github.com/CliMA/Oceananigans.jl/blob/adffd0cb67373da364a54d7f300d1f33817539bd/src/ImmersedBoundaries/mask_immersed_field.jl#L53

I'd expect it to use `solid_node` for masking scalars. Right?

For masking velocity fields for the purpose of enforcing impenetrable boundary conditions, it makes sense to me that we'd use `solid_interface`. But we should have a separate function for that.

In reality, we only _need_ masking for the velocity field for our algorithm to work correctly. However, being able to mask scalars is sometimes useful for development purposes.

@simone-silvestri can you shed some light?

**Comment by glwagner** [comment]
Maybe if we want to be defensive we should implement both

1. A `mask_immersed_cells!`" function that inserts a mask value into immersed cells. In reality this is just a convenience as noted, perhaps something to do just prior to output but not needed during the course of a computation.

2. An `enforce_impenetrability!` function that takes in the velocity field and enforces no normal flow. This is a necessary component of our time-step.

For the implementations we have now, these are identical (thus we use `solid_interface` for both), but that might not be the case in the future.

**Comment by simone-silvestri** [comment]
I agree, also we have to remember to remove the masking of the tracers as that is not required...

**Comment by francispoulin** [comment]
Maybe a silly question, but don't we want to allow for the masking of tracers?  I'm thinking about #2251, which is trying to mask a tracer.

**Comment by simone-silvestri** [comment]
We should have a way to mask the tracers for sure.

On the other hand, in our current algorithm we do not "use" immersed "halo" nodes because we to impose a no-flux condition we just set to zero the fluxes on the immersed boundaries - opposed to actual boundaries where the "no-flux" condition is imposed by mirroring the values outside the boundary and explicitly calculating a flux.

For this reason, we don't need to mask the tracer within the time integration. So, it is not necessary, but It is definitely a function we want to allow though...

**Comment by glwagner** [comment]
Right. The masking is "only" for things like plotting and visualization. We have implemented (or have tried to implement) a fully featured system for reducing fields in the presence of immersed boundaries, so we don't _need_ the mask for calculations.

On #2251 it seems we are basically using the mask to debug `is_immersed`.

**Comment by francispoulin** [comment]
Thanks for the explainations.

I like the idea of defaulting the masked regions to `NaN`.  Good for plotting and might be helpful to identify if these values are ever used, which of course they shouldn't be.

**Comment by glwagner** [comment]
this issue is a bit stale
